fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/impl/object/builder/FabricTrackedDataRegistryImpl.java

```diff
@@ -0,0 +1,160 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.object.builder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.entity.data.TrackedDataHandler;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.collection.Int2ObjectBiMap;
+
+import net.fabricmc.fabric.api.event.registry.FabricRegistryBuilder;
+import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
+import net.fabricmc.fabric.api.event.registry.RegistryIdRemapCallback;
+import net.fabricmc.fabric.mixin.object.builder.TrackedDataHandlerRegistryAccessor;
+
+public final class FabricTrackedDataRegistryImpl {
+	public static final Logger LOGGER = LoggerFactory.getLogger(FabricTrackedDataRegistryImpl.class);
+
+	private static final Identifier HANDLER_REGISTRY_ID = Identifier.of("fabric-object-builder-api-v1", "tracked_data_handler");
+	private static final RegistryKey<Registry<TrackedDataHandler<?>>> HANDLER_REGISTRY_KEY = RegistryKey.ofRegistry(HANDLER_REGISTRY_ID);
+
+	private static boolean hasStoredVanillaHandlers = false;
+	private static final List<TrackedDataHandler<?>> VANILLA_HANDLERS = new ArrayList<>();
+	@Nullable
+	private static Registry<TrackedDataHandler<?>> handlerRegistry = null;
+	private static final List<TrackedDataHandler<?>> EXTERNAL_MODDED_HANDLERS = new ArrayList<>();
+
+	private FabricTrackedDataRegistryImpl() {
+	}
+
+	public static boolean hasStoredVanillaHandlers() {
+		return hasStoredVanillaHandlers;
+	}
+
+	public static void storeVanillaHandlers() {
+		if (hasStoredVanillaHandlers) {
+			throw new IllegalStateException("Already stored vanilla handlers!");
+		}
+
+		Int2ObjectBiMap<TrackedDataHandler<?>> dataHandlers = TrackedDataHandlerRegistryAccessor.fabric_getDataHandlers();
+
+		for (TrackedDataHandler<?> handler : dataHandlers) {
+			VANILLA_HANDLERS.add(handler);
+		}
+
+		hasStoredVanillaHandlers = true;
+		LOGGER.debug("Stored {} vanilla handlers", VANILLA_HANDLERS.size());
+	}
+
+	/**
+	 * Reorders handlers in {@code TrackedDataHandlerRegistry#DATA_HANDLERS} to have a consistent order between client and server.
+	 *
+	 * <p>The order used is as follows:
+	 *
+	 * <ul>
+	 *   <li>Vanilla handlers</li>
+	 *   <li>Handlers in the Fabric API registry (sorted by ID)</li>
+	 *   <li>External modded handlers</li>
+	 * </ul>
+	*/
+	private static void reorderHandlers() {
+		Int2ObjectBiMap<TrackedDataHandler<?>> dataHandlers = TrackedDataHandlerRegistryAccessor.fabric_getDataHandlers();
```

Comment:
Question: is there a reason why this is sorted, and doesnt use the raw registry id?

This is going to have a similar issue where if the client has addional mods that arent present on the server the ids wont sync the same. Where as registry sync correctly handles this case.

Reply:
I thought it did use registry integer IDs when iterating on the registry. I assume they can be out of order or have gaps, then? If so, would it be best to add custom entries to the map using their registry integer ID + the number of vanilla handlers?

Reply:
> I thought it did use registry integer IDs when iterating on the registry. 

It does, I forgot about that. I think what I had in mind was to also add the vanilla ones, but there isnt really any benefit to doing that so what you have here is probably correct. Ill double check myself shortly.

I do wonder if the external handlers need to go before the ones in the registry? I think this approach makes the most sense as it priorities keeping the handlers tracked via the api in sync.

Reply:
I thought about this but it has the significant issue that if the amount of external handlers is not the same across the client and server, handlers added through Fabric API will start at different integer IDs and they will all be desynced.